resolver.go

```diff
@@ -373,11 +391,13 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			}()
 			err = co.WriteMsg(query)
 			if err != nil {
+				r.concurrentQueryDec()
```

Comment:
`But it looks like the decrement must be done only after co.WriteMsg() or co.ReadMsg() has been executed ?`

Yes. Between `co.WriteMsg()` and `co.ReadMsg()` is where the query is forwarded to external DNS server and the reply is accepted. In the case of concurrent calls to ServeDNS this is the block that we want to restrict to maxConcurrent instances.